pkg/cluster/minio.go

```diff
@@ -0,0 +1,213 @@
+package cluster
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/replicatedhq/ekco/pkg/util"
+	autoscalingv1 "k8s.io/api/autoscaling/v1"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+func (c *Controller) ScaleMinioStatefulset(ctx context.Context, ns string) error {
+	currentMinioSS, err := c.Config.Client.AppsV1().StatefulSets(ns).Get(ctx, "ha-minio", metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("get ha-minio statefulset: %w", err)
+	}
+
+	scaleString := currentMinioSS.Annotations["kurl.sh/desired-scale"]
+	scaleInt, err := strconv.ParseInt(scaleString, 10, 32)
+	if err != nil {
+		return fmt.Errorf("failed to decode desired scale %q: %w", scaleString, err)
+	}
+	scale32 := int32(scaleInt)
+
+	if currentMinioSS.Spec.Replicas != nil && *currentMinioSS.Spec.Replicas == scale32 {
+		return nil // already scaled
+	}
+
+	_, err = c.Config.Client.AppsV1().StatefulSets(ns).UpdateScale(
+		ctx,
+		"ha-minio",
+		&autoscalingv1.Scale{
+			Spec: autoscalingv1.ScaleSpec{
+				Replicas: scale32,
+			},
+		},
+		metav1.UpdateOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to scale ha-minio statefulset: %w", err)
+	}
+
+	return nil
+}
+
+// MigrateMinioData moves data from the un-replicated minio deployment to the HA minio statefulset,
+// using kurl's sync-object-store.
+func (c *Controller) MigrateMinioData(ctx context.Context, utilImage string, ns string) error {
+	// first, get the minio service.
+	// if it exists, we will delete it to prevent reads and writes during the migration.
+	_, err := c.Config.Client.CoreV1().Services(ns).Get(ctx, "minio", metav1.GetOptions{})
+	if err != nil {
+		if !util.IsNotFoundErr(err) {
+			return fmt.Errorf("get existing minio service: %w", err)
+		}
+	} else {
+		// delete existing minio service
+		err = c.Config.Client.CoreV1().Services(ns).Delete(ctx, "minio", metav1.DeleteOptions{})
+		if err != nil {
+			return fmt.Errorf("delete existing minio service: %w", err)
+		}
+	}
+
+	// discover the IP address of the existing minio pod to migrate from
+	podIP := ""
+	minioPods, err := c.Config.Client.CoreV1().Pods(ns).List(ctx, metav1.ListOptions{})
```

Comment:
Any reason not to use ListOptions.LabelSelector here?

Reply:
No reason in particular, I'll update (and squash commits)